File: sdk/go/flow_snapshots.go
package bunqueue

import "fmt"

func validateFlowSnapshots(
	jobs []map[string]any,
	value any,
) (map[string]map[string]any, error) {
	rawJobs, ok := value.([]any)
	if !ok || len(rawJobs) != len(jobs) {
		return nil, fmt.Errorf("invalid PUSHF response: committed job snapshots are missing")
	}

	expected := make(map[string]string, len(jobs))
	for _, job := range jobs {
		expected[asString(job["id"])] = asString(job["queue"])
	}
	snapshots := make(map[string]map[string]any, len(rawJobs))
	for _, item := range rawJobs {
		raw, ok := item.(map[string]any)
		if !ok {
			return nil, fmt.Errorf("invalid PUSHF response: committed job snapshot is malformed")
		}
		id, ok := raw["id"].(string)
		queue, expectedID := expected[id]
		if !ok || !expectedID || snapshots[id] != nil || raw["queue"] != queue {
			return nil, fmt.Errorf(
				"invalid PUSHF response: committed job IDs or queues do not match request",
			)
		}
		snapshots[id] = raw
		delete(expected, id)
	}
	if len(expected) != 0 {
		return nil, fmt.Errorf(
			"invalid PUSHF response: committed job IDs or queues do not match request",
		)
	}
	return snapshots, nil
}

File: sdk/elixir/lib/bunqueue/flow_snapshots.ex
defmodule Bunqueue.FlowSnapshots do
  @moduledoc false

  alias Bunqueue.ProtocolError

  @spec index!([map()], term()) :: %{String.t() => map()}
  def index!(jobs, snapshots) when is_list(snapshots) and length(jobs) == length(snapshots) do
    expected = Map.new(jobs, &{&1["id"], &1["queue"]})

    by_id =
      Enum.reduce(snapshots, %{}, fn
        snapshot, result when is_map(snapshot) ->
          id = snapshot["id"]
          queue = snapshot["queue"]

          if not is_binary(id) or Map.get(expected, id) != queue or Map.has_key?(result, id) do
            invalid_ids!()
          end

          Map.put(result, id, snapshot)

        _snapshot, _result ->
          raise ProtocolError, message: "Invalid PUSHF response: job snapshot is invalid"
      end)

    if map_size(by_id) != map_size(expected), do: invalid_ids!()
    by_id
  end

  def index!(_jobs, _snapshots) do
    raise ProtocolError, message: "Invalid PUSHF response: committed job snapshots are missing"
  end

  defp invalid_ids! do
    raise ProtocolError,
      message: "Invalid PUSHF response: committed job IDs do not match the request"
  end
end
